Support Google Container Registry

petemounce opened the pull request:
Not sure whether this will be a welcome addition (no reply inside #24), but I decided to make a start since I'll be able to use this via our fork either way.

### What's going on here:

* Extracted the ECR-specific functions to the ecr-registry-provider.bash include.
  * They're verbatim implementations, but a couple had their names changed since there'll be GCR-specific implementations so I made the names generalised.
* Introduced the provider-specific include at the top of `hooks/pre-command`. This will fail with an error if the plugin is misconfigured.

Initial commit is _just this_. I'm still interested to see whether the maintainers (I assume @zsims ?) are interested in this addition.

### Still to-do (whether that's this PR or followups)

* [x] docs updates
* [x] GCR implementation
    * [x] BATS for GCR implementation

Commit:
Support GCR step 1; factor to library

Commit:
more refactor - GCR has no concept of repository for images nor therefore max-age (those are done via GCS TTL policy)

Commit:
tweaks

Commit:
GCR WIP

zsims commented:
Hey, thanks for the PR! Sorry for no reply on #24 -- has been a bit crazy :S

I think this does make sense in the same plugin. I was originally thinking it would be easier to have a separate plugin, but with the same interface. However, one thing to keep in mind would be testing -- we don't use GCR so we'd have to rely on the bats tests.

zsims commented on hooks/lib/ecr-registry-provider.bash:
```diff
@@ -0,0 +1,106 @@
+      # Handle plain key=value, e.g
+      # ecr-tags:
+      #   KEY_NAME: 'key-value'
+      key_name=$(echo "${name}" | sed 's/^BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ECR_TAGS_//')
```
reckon it's worth "standardising" on the interface, e.g. this would become `tags` rather than `ecr-tags`?

zsims replied:
we'd probably want both for backwards compat, but that's easy :-)

petemounce replied:
Actually - I don't know whether GCR has an equivalent to this ECR concept. I'm unfamiliar with ECR in enough depth to know what to look for, to be honest!

zsims replied:
In that case, reckon it's worth splitting this into another plugin? e.g. `docker-gcr-cache-buildkite-plugin`? We can align on the interface fairly easily if people want to switch between them

I'm just wondering how easy future updates are going to be given we don't use GCR, and you don't use ECR :)

petemounce replied:
The closest GCR has is the concept of labels, which get applied to the GCR itself (which actually turns into a user-visible GCS bucket). It's not possible to label images/image-families/channels within the GCR.

We do use ECR; we operate with AWS within the borders of China, where Google doesn't operate. We have a set of issues with traffic that needs to cross the Great Firewall of China So, I might not use ECR day to day, but I'd prefer to have a single plugin with multiple implementations within it. One of the things I'd like to be able to do is push to a number of registries, for example.

zsims replied:
Makes sense, good reasons :)

zsims commented on hooks/lib/gcr-registry-provider.bash:
```diff
@@ -0,0 +1,25 @@
+login() {
+  # Currently assume the use of docker-credential-gcr to manage AuthN transparently.
```
good idea. We could probably do this for ECR too, however it would mean consumers have to "do more" to use the plugin

petemounce replied:
Yes, definitely; I'll flag that in the readme but also not address it immediately. I'm just getting to the point internally to actually _try_ this for the first time, it has been a hefty week.

zsims replied:
👍 we can certainly tackle in a new PR.

zsims commented on hooks/lib/stdlib.bash:
```diff
@@ -0,0 +1,9 @@
+log_fatal() {
+  echo "${@}"
```
```suggestion
  echo "${@}" 1>&2
```

worth putting on stderr

zsims reviewed (commented):
Looks great, would be happy to accept this with the tests. I reckon more overlap/same interface the better

Commit:
Update hooks/lib/stdlib.bash

Co-authored-by: Zac Sims <zsims@users.noreply.github.com>

Commit:
debug issue on windows, suspect missing context value with non-standard dockerfile location, OR file path separators...

Commit:
fail-faster with error message

Commit:
More debug

Commit:
refactor, show where cwd is

Commit:
doh

Commit:
argh

Commit:
weird?

Commit:
wat.

Commit:
uh

Commit:
really?

Commit:
file

Commit:
quotes

Commit:
'

Commit:
really

Commit:
sigh

petemounce commented:
(This commit stream is me iterating a docker build on a Windows machine, and I'm having a great time, clearly)

Commit:
works internally. It's always quoting, always....

Commit:
docs, plugin definition, more refinement

Commit:
oops, err

zsims commented on hooks/pre-command:
```diff
@@ -121,16 +33,18 @@ read_list_property() {
 
   echoerr 'DOCKERFILE'
-  echoerr "+ ${docker_file}:${target:-<target>}"
+  echoerr "+ ${docker_file}:${target:-"<target>"}"
```
Are these extra quotes required?

petemounce replied:
No, that's muscle memory on my part. I tend to always enclose strings in quotes and variables inside {} because it's easier than remembering where it will be necessary and where it won't be necessary, compared to just reflexively doing it everywhere.

petemounce committed:
Added BATS coverage for stdlib and GCR-specific function implementations

petemounce committed:
logic coverage 1

petemounce commented:
@zsims please take another look? I've added test coverage, and intend to add 2 more to `pre-command.bats`.

petemounce committed:
Remainder of coverage

petemounce commented:
@zsims Please also see #27?

I reckon this is now ready for re-review. Since last time, I have
* moved the functions from `hooks/pre-command` to `hooks/lib/stdlib.bash` to move them out the way. At the time, I'd hoped it was possible to stub out bash functions, but I later didn't find out how, so ...
* I added `hooks/lib/stub-registry-provider.bash` which makes stubs for things so that the `tests/pre-command.bats` didn't have to repeat...
* ... the test coverage I wrote in `tests/stdlib.bash`
* added test coverage `tests/gcr-registry-provider.bats`

That takes us from 2 tests to 17. I never knew it was so comparatively straightforward to write bash test coverage, so I'm really grateful for the lesson!

In terms of backwards compatibility, is there anything you'd like me to do (though, probably if so, in a follow-up - this PR has ballooned!). Things I thought of:

* make `additional-build-args` an array instead of a string, since that simplifies that bit of script that glues that onto the end of the `docker build` down to `"${BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ADDITIONAL_BUILD_ARGS[@]}"`
* add `.pre-commit-config.yaml` with `shellcheck` and `shfmt` and some other quality-of-life stuff (see pre-commit.com)
* make the `target` be an arg to `compute_tag` instead of leaking in from outer scope
* Add a plugin property `image-name` to replace `ecr-name`
  * ... but maintain backwards compatibility by coalescing back to `ecr-name`'s env var
* Maybe talk about renaming the repository to `docker-cache-buildkite-plugin`?
  * That has some entertaining injection security implications, however - github will redirect requests both UI and clone to the new name from the old name. However, if someone makes a new replacement `docker-ecr-cache-buildkite-plugin` with the same tags as here, that will silently man-in-the-middle peoples' pipelines.
  * perhaps @lox or @toolmantim might like to give some advice on what is best to do here?

petemounce commented:
(The tests run and pass on Windows and macOS hosts)

toolmantim commented:
Sounds like a rename is a good idea!

If you trust seek-oss enough to not use SHA-pinned plugins, then perhaps it's probably okay to let the name sit there available for another repo to steal back at some point?

petemounce commented:
I'm a very trusting sort. I'd actually forgotten that it's an commercial org :)

toolmantim commented:
😁

72636c commented:
Hey, thanks for this @petemounce! @zsims has subbed out but I will try to review soon 🙏

RE: the rename, I think the least surprising way would be to fork this repo then archive the current one.

72636c commented on hooks/lib/stdlib.bash:
```diff
@@ -0,0 +1,88 @@
+read_build_args() {
+  read_list_property 'BUILD_ARGS'
+  for arg in ${result[@]+"${result[@]}"}; do
+    build_args+=("--build-arg=${arg}")
```
Was there something wrong with the old version here? It's NBD for build args but I'm generally wary of escaping with this syntax.

```suggestion
    build_args+=('--build-arg' "${arg}")
```

72636c replied:
Ah ignore this, arg parsers seem to do a good job handling this case

72636c commented on hooks/lib/stdlib.bash:
```diff
@@ -0,0 +1,88 @@
+  echoerr "${@}"
+  # use the last argument as the exit code
+  exit_code="${*: -1}"
+  if [[ "${exit_code}" =~ ^[\d]+$ ]]; then
```
This should be more portable:

```suggestion
  if [[ "${exit_code}" =~ ^[0-9]+$ ]]; then
```

72636c commented on hooks/pre-command:
```diff
@@ -1,202 +1,50 @@
-  "${context}" || exit 1
+  image_build_args=(
+    "build"
+    "--file=${docker_file}"
```
Same deal with these flags. It's less a matter of security since you shouldn't be running pipelines that you don't trust on your agents, but I think there's potential for surprising behaviour like:

```
docker build ---file=My Documents\Dockerfile
```

Can we revert to the previous approach?

72636c commented on hooks/lib/gcr-registry-provider.bash:
```diff
@@ -0,0 +1,20 @@
+    echoerr "registry-hostname had no value, defaulting to gcr.io"
+    BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_REGISTRY_HOSTNAME="gcr.io"
+  fi
+  echo "${BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_REGISTRY_HOSTNAME}/${BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_GCP_PROJECT}/${BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ECR_NAME:-"$(get_default_image_name)"}"
```
Ugh. Not for this PR, but we should definitely have a generic alias for this option as you've suggested.

Note that the README says that this is an ECR-specific option, so we should either update it _or_ drop this logic from here until we have that `image-name`.

72636c replied:
Let me know which way you'd prefer to go on this one ^ @petemounce

Apart from that, looks all ready to merge!

petemounce replied:
I think I'd prefer to leave this as-is within this PR, and iterate afterwards.

petemounce replied:
Actually, wait - I'm not sure I follow what you're suggesting. Please would you elaborate?

GCR URIs take the form `https://{host}/{gcp-project}/{container}:{tag} - so, actually, I think `container` is the equivalent to ECR's repository now that I think about this more freshly.

72636c replied:
> I think container is the equivalent to ECR's repository

Yep, looks that way to me too. Happy to iterate on it.

Above I was thinking that we should document that `ecr-name` affects the GCR configuration, but it's probably not necessary given it's optional and we'd probably want to fix up the name in future.

72636c commented on hooks/pre-command:
```diff
@@ -1,202 +1,50 @@
+  # We can't quote BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ADDITIONAL_BUILD_ARGS, because it's passed here as a string instead of a bash array.
+  # shellcheck disable=SC2086
+  docker "${image_build_args[@]}" ${BUILDKITE_PLUGIN_DOCKER_ECR_CACHE_ADDITIONAL_BUILD_ARGS:-} "${context}" ||
+    log_fatal "^^^ +++" 1
```
Appreciate all these little improvements 👍

72636c reviewed (commented):
Thanks for your patience! It's been pretty full-on here over the past few weeks. This is a massive improvement and the tests are awesome 🙏

Just have a few nits that should be easily resolved. Let me know if you're busy and I can run with it.

Commit:
better regex match choice

Co-authored-by: Ryan Ling <ryan@outlook.com.au>
